### DISCUSSION TITLE
Improving Type-checking

### DISCUSSION THREAD

**Discussion by balacij** [JacquesCarette/Drasil#3788] [Q&A] [answered]
In our last meeting, @JacquesCarette mentioned that we should give priority to cleaning up things related to the type checker. So, let's discuss and confirm the nature of our work and what kind of changes we can/should make needed to “clean up” the type checker. In other words, let's figure out what “clean up” means :wink: .

## Where it's at

Quick recap: 
|  Language | Typed? |
| --------: | :----- |
|      Expr | Yes    |
| ModelExpr | No     |
|  CodeExpr | No     |

### Design choices

Recalling a few of our design choices:
1. The type-checker does not piggyback on Haskell's type system, building our own type-checker that checks whole expressions. We wanted to permit ill-typed expressions in these languages, in specific scenarios, and we also ended up preferring to stray away from GADT because of two reasons (IIRC):
   1. We were (at the time) reluctant to add a type argument to QuantityDict to carry type information, and,
   2. even if we preferred that, we could not easily do that because limitations of the current ChunkDB design (unblocked by #2873).
2. We split our previous monolithic `Expr` into the 3 flavours so that we can restrict certain terms and features to specific areas of code.
   * Expr is for simpler mathematics, expressions with a definite value.
   * CodeExpr is an extension of Expr with extra features for OO.
   * ModelExpr is an extension of Expr with all the goodies we normally use in pure mathematics and analysis (well, not _all_, but more than Expr).
3. We opted to go for a bidirectional type-checker design. The current type-checker we have is rather uncomplicated as well, as it only supports Expr.

Unfortunately, we only got to typing Expr, but now we want to extend it to type checking the others, whilst minimizing code duplication, amongst other things.

### How the current type-checker works

The current type-checker is built around the below code:

https://github.com/JacquesCarette/Drasil/blob/85cf74daaccb74131373df9d52cc2bc751fd48e9/code/drasil-lang/lib/Language/Drasil/WellTyped.hs#L10-L38

In the above code, we see:
1. A “type error” type (a String), which is currently used (attempted) to build a tree view of expressions for when type-checking fails, so that we can print out meaningful error messages about where issues occurred, at the AST level.
2. A “typing context” type (a Map of UIDs to types), which is used to provide the environment of available variables to the type-checker, and for the type-checker to manipulate (~semi-writable -- overwrites only permitted for appended types) as needed.
3. A “typed” typeclass, which is used to indicate that some_thing_ can be typed against a specific type system/language.

Recalling the code related to Expr, we have a type-checker that relies purely on simple inference:

https://github.com/JacquesCarette/Drasil/blob/85cf74daaccb74131373df9d52cc2bc751fd48e9/code/drasil-lang/lib/Language/Drasil/Expr/Lang.hs#L232-L428

### Issues with the current work

We see:
1. The code is very verbose and hard to read.
2. Error messages are somewhat okay, but lack the majority of the contextual information for users to debug where a type error occurred.
3. A lot of code duplication.

Ultimately, there's a lot of room for improvement.

## What we want

First, we want the inverse of the 3 issues:
1. Easier to read, shorter code.
2. Better type error messages.
3. Less code duplication.

Afterwards, we want to extend the type-checker to ModelExpr and CodeExpr, while being mindful of code duplication. ModelExpr will be a bit more complicated than extension to CodeExpr because of a non-trivial interest: type variables.

## Where should we go?

For now, we will focus on the earlier 3 issues.

To address the first issue and the parts of the third, I think we can switch over to a monadic type inference system. Right now, monads are severely underused. We can review how much code duplication is left over after switching to a monadic style.

Now, for improved type error messages, I'm not completely sure. One thing is for sure: we need to be able to track where we are in the expression as we go along. Do we do that from “source”? I don't see any easy way to do that without writing a GHC plugin, but I'm not sure if that's the route we want to go down. So, should we track just the AST and print out a horizontal tree that prints the AST with the root in the right-most position? Should we just try to print out Haskell-like code? I'm not quite sure.

**Comment by smiths** [comment]
Nice summary @balacij.  I don't know the answers to your questions, but @JacquesCarette very likely will have some thoughts.  Maybe we should discuss this during our morning meeting on Monday?

**Reply by balacij** [reply, marked as answer]
Thank you! Hmm, I don't think it would hurt, but I think this one can be done over GitHub, so perhaps as a lower priority item?

**Comment by JacquesCarette** [comment]
As discussed today, we also want the code to "look like" the typing rules as much as possible.

One of the things to do to clean up the code a lot is to move the error-generating code "elsewhere". And error happens when a **conflict** arises between reality (the `Expr`) and one of the rules (in context). So given the inputs to `check` (or `infer`) and the *detected conflict* then a completely separate routine should be able to reconstruct the error.

Then the remaining work becomes one of 'classifying' the different kinds of conflicts that can arise from the rules.

I wonder if there's something we can learn from Andrej Bauer's PL zoo regarding this.
